### ISSUE TITLE
ruff does not format the code after ctrl+alt+l (almost always)

### ISSUE THREAD

**Issue by chaush-server** [koxudaxi/ruff-pycharm-plugin#531]
When you press ctrl alt L or in other words, launch code reformat even with a button from the menu, the code is not formatted, almost always (if it is, I return everything back ctrl + z several times and the bug manifests itself). I found a specific behavior when I quickly press ctrl + alt + L several times in a row, after that the code is formatted. And if you format it with expectation, it is not formatted. Also, I have assigned ruff to alt shift f and by this combination the formatting is triggered every time. I also tried all ruff settings (global ruf, lsp), turned off all third-party plugins

 - IDE: PyCharm Pro 2024.3.1
 - OS: Win 11
 - Ruff Version 0.8.1
 - Plugin version 0.0.42 

**Screenshots**
![изображение](https://github.com/user-attachments/assets/251bfeec-9cd0-4b2b-9f6c-e59493ea726c)
![изображение](https://github.com/user-attachments/assets/efa36f48-fc8f-4960-b710-06b13ab4e05e)

**Comment by rsheftel** [comment]
I have run into the same issue with plugin verion 0.0.42 and ruff version 0.7.1. To get the ruff plugin to format the code I need to press Ctrl-Alt-L at least 3 times in succession. There are no warnings or errors, but that is the only way to make it format.

**Comment by jaens** [comment]
Had the same problem, the "use ruff format (experimental)" setting fixed it for me, at least temporarily:
![image](https://github.com/user-attachments/assets/78b14853-b444-488f-bd29-4627c37580f8)

**Comment by Olegt0rr** [comment]
Got the same issue

```
ruff==0.9.1
```

```
PyCharm 2024.3.1.1 (Professional Edition)
Build #PY-243.22562.220, built on December 18, 2024
Runtime version: 21.0.5+8-b631.28 aarch64 (JCEF 122.1.9)
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Toolkit: sun.lwawt.macosx.LWCToolkit
macOS 15.2
GC: G1 Young Generation, G1 Concurrent GC, G1 Old Generation

Non-Bundled Plugins:
  com.koxudaxi.ruff (0.0.42)
```

**Comment by koxudaxi** [comment]
Sorry for the late confirmation.
I have tested formatting in my environment and it works with a single press.
Here is my environment:

<img width="940" alt="image" src="https://github.com/user-attachments/assets/ea2be0e2-f415-408f-b2fc-fbe4a49c027f" />


IDE: PyCharm Pro 2024.3.1 #PY-243.22562.220
OS: macOS 15.1.1
Ruff Version 0.9.1 (12f86f39a 2025-01-10)
Plugin version 0.0.42

As @jaens pointed out, the `Use ruff format...` option needs to be enabled for formatting to work. If formatting worked in previous versions without this option enabled, that would be considered a bug.

**Comment by koxudaxi** [comment]
We initially treated format as an experimental option, but if this is going to be used by everyone, it might be better to turn it on by default.

Please comment if the formatting does not take place even after enabling this format option. I also have a windows environment, so I will try that as well.

**Comment by Olegt0rr** [comment]
Formatting is important for me. It would be nice for me to make it enabled by default! :)

**Comment by koxudaxi** [comment]
I also tested it in windows. It works fine.

PyCharm 2024.3.1.1 (Professional Edition)
Build #PY-243.22562.220, built on December 18, 2024
Runtime version: 21.0.5+8-b631.28 amd64 (JCEF 122.1.9)
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Toolkit: sun.awt.windows.WToolkit
Windows 11.0
GC: G1 Young Generation, G1 Concurrent GC, G1 Old Generation
Memory: 2048M
Cores: 20
Registry:
  ide.experimental.ui=true
  i18n.locale=
Non-Bundled Plugins:
  com.redhat.devtools.lsp4ij (0.9.0)
  com.intellij.ml.llm (243.22562.251.6)
  com.koxudaxi.ruff (0.0.42)

**Comment by koxudaxi** [comment]
@Olegt0rr
I used the code for the test.
```python
def foo(
    bar: int,
        baz: str,
):
    eggs = [
        1,
            2,
    ]
```
after enter `alt + command + L`
```python
def foo(
    bar: int,
    baz: str,
):
    eggs = [
        1,
        2,
    ]
```

The project doesn't have the any ruff config file like `pyproject.toml` or `ruff.toml`





```
Non-Bundled Plugins:
  com.koxudaxi.ruff (0.0.42)
```


Which do you use the LSP Client?
Intellij's LSP client or `LSP4IJ` ?

<img width="517" alt="image" src="https://github.com/user-attachments/assets/410b082c-6d79-4771-a84d-54832e085f08" />

if you select LSP4IJ when you should install LSP4IJ as a plugin.
<img width="1229" alt="image" src="https://github.com/user-attachments/assets/ab4b72c8-80ad-496d-aafc-b468bbf99cbb" />

**Comment by koxudaxi** [comment]
@Olegt0rr 
Thank you for providing the video.
One concerning point is that "Running Ruff Formatter" is not showing in the Window progress bar at the bottom after you execute Reformat. In your video, "Reformat Code" appears briefly, but it should be followed by "Running Ruff Formatter".
For confirmation, are you using any other plugins or formatters? During the development of this plugin, I noticed that only one AsyncDocumentFormatter can run at a time, so if another formatter is executing, the Ruff formatter won't work.


<img width="265" alt="image" src="https://github.com/user-attachments/assets/f6fb8a07-478c-49d6-ab1a-ef4b7528ee72" />

I expect the message after that.
<img width="452" alt="image" src="https://github.com/user-attachments/assets/c52f1ee5-4d53-4e43-b74c-e7a27596479f" />

**Comment by Olegt0rr** [comment]
I'm not using any formatters...

But just removed all old plugins from IDE and... magic! ruff works

Unfortunately, in my emotions I deleted all plugins in a row, both the ones that were turned off and the ones I haven't used in a long time... And now I can't say which one had compatibility problems (which would be useful for those who are experiencing problems)

**Comment by tuialexandre** [comment]
I'm having similar issue on a fresh install of PyCharm. I was actually testing the IDE but added the Ruff plugin, as I'm used to it's extension in VSCode. I tried the "Use ruff format (experimental)" option to no success.

PyCharm 2024.3.1.1 (Community Edition)
Ruff plugin: 0.0.42

**Comment by mjakl** [comment]
I've also have/had the issue with needing to run formatting multiple times. It looks like ruff and PyCharm formatting are working both and there's a race condition. Sometimes ruff wins, sometimes the other wins. I don't know why this happens.

My solution I set a hotkey to "Run ruff" to Cmd-Option-L  and use that. This works well and isn't a big nuisance. It would of course be nice if "Run ruff on save" would work.

(I've tried LSP4IJ but it crashes, so I'm using IntelliJ's LSP client)

👍  for "Use ruff format" as default.

**Comment by angelozerr** [comment]
> I've also have/had the issue with needing to run formatting multiple times. It looks like ruff and PyCharm formatting are working both and there's a race condition. Sometimes ruff wins, sometimes the other wins. I don't know why this happens. It' also much more likely that ruff works if I save the file very quickly after an edit. If I wait a bit longer, saving doesn't trigger ruff anymore.
> 
> My solution I set a hotkey to "Run ruff" to Cmd-Option-L  and use that. This works well and isn't a big nuisance. It would of course be nice if "Run ruff on save" would work.
> 
> (I've tried LSP4IJ but it crashes, so I'm using IntelliJ's LSP client)

Please create an issue in LSP4IJ 
> 
> 👍  for "Use ruff format" as default.

**Comment by mjakl** [comment]
> Black settings are in `Tools > Black`. Try enabling, testing, then disabling it and let us know if that works for you.

Thanks for the suggestion. I tried it, but could not reproduce your results. What I did:

- Enabled black on Reformat and on save - didn't work (no reformatting at all)
- Disabled ruff on Reformat and on save (kept black on) - black did the formatting
- Disabled black and enabled ruff - ... this is where it gets interesting:

When I edited the file and **quickly** pressed save, the reformat worked! When I waited a bit, I noticed a process in my status bar (too quick to see what it is), after that it didn't work anymore - this "thing" happened about 1 second after edits.

I did a screen recording to see what pops up: Pyright!

**I disabled the Pyright plugin in PyCharm, now ruff works flawlessly!**

So, my solution to get ruff working on Save and on Reformat is to disable the Pyright plugin.

(Note: My Pyright was set to "Running mode" "Command line"; it seems **ruff is also working when I set the Pyright "Running mode" to "Native LSP client"**).

HTH!

**Comment by qiuxiaomu** [comment]
> Sorry for the late confirmation. I have tested formatting in my environment and it works with a single press. Here is my environment:
> 
> <img alt="image" width="940" src="https://private-user-images.githubusercontent.com/630670/402969741-ea2be0e2-f415-408f-b2fc-fbe4a49c027f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg1NzczMjcsIm5iZiI6MTczODU3NzAyNywicGF0aCI6Ii82MzA2NzAvNDAyOTY5NzQxLWVhMmJlMGUyLWY0MTUtNDA4Zi1iMmZjLWZiZTRhNDljMDI3Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwM1QxMDAzNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NTEyMmFjNTI0ZmFmMDEwMTUwNzE3ZWVlMzU4ODg5N2E0ZGY2YzNlNzg1MzYzMDVhNDE3YzA3Y2IxY2RkYmQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.C5Lwr1e-jMU5zOqYtcZGiUOiqAHf-n29TXYoaBQi8w4">
> IDE: PyCharm Pro 2024.3.1 #PY-243.22562.220 OS: macOS 15.1.1 Ruff Version 0.9.1 (12f86f39a 2025-01-10) Plugin version 0.0.42
> 
> As [@jaens](https://github.com/jaens) pointed out, the `Use ruff format...` option needs to be enabled for formatting to work. If formatting worked in previous versions without this option enabled, that would be considered a bug.

With this setting it works for me with the keyboard shortcut; it however interferes with format on save, if we tick the first "run ruff when the python file is save".

**Comment by koxudaxi** [comment]
I apologize for inconvenience. Currently, I am working on modification to move operation including format to LSP.
If this implementation succeeds, I think it will solve problems of interference with different formatters and not supporting partial formatting.
Sorry, but please wait a little more time.
https://github.com/koxudaxi/ruff-pycharm-plugin/pull/550

**Comment by koxudaxi** [comment]
Sorry for the delay. I have released version 0.0.45
https://plugins.jetbrains.com/plugin/20574-ruff/versions/stable/691739
I've made improvements to the Reformat functionality and strengthened LSP support.
I believe this resolves several potential issues.
Everything seems to be working well in my testing, but please let me know if you encounter any problems.

@angelozerr
It might be better to create an issue in the LSP4IJ repository, but I'll also document it here.
I wanted to fully migrate to LSP, but I noticed a particular problem.
There's no issue with formatting, but this plugin also implements CodeAction's FixAll as part of the Reformat process.
Of course, there's no way to hook FixAll into the LSP Formatting Feature, so
I experimentally implemented a formatter that sequentially performs format, lint FixAll, and Organize Imports commands.
Is this approach correct?
Is there a better way to execute a sequence of commands?

(I need to investigate whether JetBrains' LSP functionality does the same thing, so I'm not sure if this PR will be merged.)

However, because of this issue, I'm uncertain whether both can continue to be supported. I'll need to take some time to investigate this matter.
https://github.com/redhat-developer/lsp4ij/issues/459

**Comment by angelozerr** [comment]
> Sorry for the delay. I have released version 0.0.45
> https://plugins.jetbrains.com/plugin/20574-ruff/versions/stable/691739
> I've made improvements to the Reformat functionality and strengthened LSP support.
> I believe this resolves several potential issues.
> Everything seems to be working well in my testing, but please let me know if you encounter any problems.
> 
> @angelozerr
> It might be better to create an issue in the LSP4IJ repository, but I'll also document it here.


Please create an issue by explaing in detail what you need 1s api, or better please create a pr with the required api.

> I wanted to fully migrate to LSP, but I noticed a particular problem.
> There's no issue with formatting, but this plugin also implements CodeAction's FixAll as part of the Reformat process.
> Of course, there's no way to hook FixAll into the LSP Formatting Feature, so
> I experimentally implemented a formatter that sequentially performs format, lint FixAll, and Organize Imports commands.
> Is this approach correct?
> Is there a better way to execute a sequence of commands?
> 
> (I need to investigate whether JetBrains' LSP functionality does the same thing, so I'm not sure if this PR will be merged.)
> 
> However, because of this issue, I'm uncertain whether both can continue to be supported. I'll need to take some time to investigate this matter.
> https://github.com/redhat-developer/lsp4ij/issues/459

**Comment by arjun-menon** [comment]
Got the same issue, but resolved by setting the "use experimental" flag. Shouldn't this be default behavior now, considering this thread was opened in November 2024?
